ocaml/xapi/stream_vdi.ml

```diff
@@ -100,6 +100,64 @@ let write_block ~__context filename buffer ofd len =
     then raise (Api_errors.Server_error (Api_errors.client_error, [ExnHelper.string_of_exn e]))
     else raise e
 
+type descriptor = {
+  length : int64;
+  status_flags : int32;
+}
+
+let flag_hole = 1l
+let flag_zero = 2l
+
+let cycle_descriptors descriptor_list offset =
+  let rec range acc a b =
+    if a < b then acc
+    else range (a::acc) (Int64.add a Int64.minus_one) b
+  in
+
+  let is_empty e =
+    let has_flag flag =
+      Int32.logand e.status_flags flag = flag
+    in
+    (* We assume the destination is prezeroed, so we do not have to copy zeroed extents *)
+    (has_flag flag_hole) || (has_flag flag_zero)
+  in
+
+  let find_blocks descriptor offset increment =
+    match (is_empty descriptor) with
+      | false -> begin
+        let start_chunk = Int64.div offset increment in
+        (* If the chunk ends at the boundary don't include the next chunk *)
+        let end_chunk = Int64.div (Int64.add (Int64.add descriptor.length  offset) Int64.minus_one ) increment in
+        let chunks = range [] end_chunk start_chunk in
+        chunks, descriptor.length
+        end
+      | true -> [], descriptor.length
+  in
+  (* Only compare to most recent chunk added*)
+  let add_new a b =
+    match b with
+    | [] -> a::b
+    | _ -> begin
```

Comment:
Why not 
```ocaml
| hd :: _ when Int64.equal a hd -> b
| _ -> a::b
```

Reply:
See http://dev.realworldocaml.org/lists-and-patterns.html#scrollNav-6 for an explanation of the when close: it can add extra conditions to a pattern